### PR TITLE
Improve the ymodem code

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
Improve the ymodem code, and modify the ymodem protocol, add a new header, and improve the transmission speed
## Impact
ymodem
## Testing
stm32 & esp32s3 use ymodem

### CODE REVIEW

File: system/ymodem/rb_main.c [hunk 1]
```diff
@@ -106,20 +107,23 @@ static int handler(FAR struct ymodem_ctx *ctx)
...
-      else if (ret < size)
-        {
-          return ERROR;
+          ret = write(ym_fd->file_fd, ctx->data + i, size - i);
+          if (ret < 0)
+            {
+              return ret;
+            }
+
+          i += ret;
```
Comment (pkarashchenko) [inline system/ymodem/rb_main.c:119]:
```suggestion
            }
          else
            {
              i += ret;
            }
```

File: system/ymodem/rb_main.c [hunk 2]
```diff
@@ -176,40 +183,74 @@ int main(int argc, FAR char *argv[])
...
 
+  memset(ctx, 0, sizeof(struct ymodem_ctx));
+  ctx->packet_handler = handler;
+  ctx->customsize = customsize;
+  ctx->timeout = timeout;
+  ctx->priv = &ym_fd;
+  if (recvfd)
+    {
+      ctx->recvfd = recvfd;
+      ctx->sendfd = recvfd;
```
Comment (pkarashchenko) [inline system/ymodem/rb_main.c:233]: So both `recvfd`and `sendfd` are set to the same value? What is the point with that?

File: system/ymodem/rb_main.c [hunk 3]
```diff
@@ -176,40 +183,74 @@ int main(int argc, FAR char *argv[])
...
+  ctx = malloc(sizeof(*ctx));
+  if (ctx == NULL)
     {
-      close(ctx.recvfd);
+      fprintf(stderr, "ERROR: can't malloc ymodem ctx\n");
+      ret = -ENOMEM;
+      goto out;
     }
 
+  memset(ctx, 0, sizeof(struct ymodem_ctx));
```
Comment (pkarashchenko) [inline system/ymodem/rb_main.c:225]: can `calloc` and remove `memset`

File: system/ymodem/sb_main.c [hunk 4]
```diff
@@ -108,55 +131,84 @@ static void show_usage(FAR const char *progname, int errcode)
 int main(int argc, FAR char *argv[])
 {
   struct ymodem_fd ym_fd;
-  struct ymodem_ctx ctx;
+  struct ymodem_ctx *ctx;
```
Comment (pkarashchenko) [inline system/ymodem/sb_main.c:134]:
```suggestion
  FAR struct ymodem_ctx *ctx;
```

File: system/ymodem/ymodem.h [hunk 5]
```diff
@@ -55,7 +55,8 @@ struct ymodem_ctx
 {
   uint8_t header;
   uint8_t seq[2];
-  uint8_t data[YMODEM_PACKET_1K_SIZE];
+  uint8_t *data;
```
Comment (pkarashchenko) [inline system/ymodem/ymodem.h:58]:
```suggestion
  FAR uint8_t *data;
```

**Comment by xiaoxiang781216** [comment]
fixed here: https://github.com/apache/nuttx-apps/pull/1720

**Pull request closed without merging**